logger/main.go

```diff
@@ -42,26 +44,49 @@ func main() {
 
 	client := etcd.NewClient([]string{"http://" + publishHost + ":" + publishPort})
 
-	// Wait for terminating signal
-	exitChan := make(chan os.Signal, 2)
+	signalChan := make(chan os.Signal, 1)
+	drainRespChan := make(chan *etcd.Response)
+	drainChan := make(chan string)
+	stopChan := make(chan bool)
+	exitChan := make(chan bool)
 	cleanupChan := make(chan bool)
-	signal.Notify(exitChan, syscall.SIGTERM, syscall.SIGINT)
+	signal.Notify(signalChan, syscall.SIGTERM, syscall.SIGINT)
 
-	go syslogd.Listen(exitChan, cleanupChan, fmt.Sprintf("%s:%d", logAddr, logPort))
+	// ensure the drain key exists in etcd.
+	if _, err := client.Get(publishPath+"/drain", false, false); err != nil {
+		setEtcd(client, publishPath+"/drain", drainURI, 0)
+	}
 
+	go client.Watch(publishPath+"/drain", 0, false, drainRespChan, stopChan)
+	go syslogd.Listen(exitChan, cleanupChan, drainChan, fmt.Sprintf("%s:%d", logAddr, logPort))
 	if enablePublish {
-		go publishService(client, publishHost, publishPath, strconv.Itoa(logPort), uint64(time.Duration(publishTTL).Seconds()))
+		go publishService(exitChan, client, publishHost, publishPath, strconv.Itoa(logPort), uint64(time.Duration(publishTTL).Seconds()))
 	}
 
-	// Wait for the proper shutdown of the syslog server before exit
-	<-cleanupChan
+	for {
+		select {
+		case er := <-drainRespChan:
+			drainChan <- er.Node.Value
+		case <-signalChan:
+			close(exitChan)
+			stopChan <- true
+		case <-cleanupChan:
+			return
+		}
+	}
 }
 
-func publishService(client *etcd.Client, host string, etcdPath string, port string, ttl uint64) {
+func publishService(exitChan chan bool, client *etcd.Client, host string, etcdPath string, port string, ttl uint64) {
+	t := time.NewTicker(time.Duration(publishInterval))
+
 	for {
-		setEtcd(client, etcdPath+"/host", host, ttl)
-		setEtcd(client, etcdPath+"/port", port, ttl)
-		time.Sleep(time.Duration(publishInterval))
+		select {
+		case <-t.C:
+			setEtcd(client, etcdPath+"/host", host, ttl)
+			setEtcd(client, etcdPath+"/port", port, ttl)
+		case <-exitChan:
```

Comment:
actually since this isn't part of the PR let's just add another TODO here as well, if you wouldn't mind